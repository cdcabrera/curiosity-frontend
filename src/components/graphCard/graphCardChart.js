import React from 'react';
import PropTypes from 'prop-types';
import { Card, CardActions, CardBody, CardHeader, CardTitle, Title } from '@patternfly/react-core';
import { useProduct, useProductGraphTallyQuery } from '../productView/productViewContext';
import { useGraphCardContext, useGetTally, useGetMetrics } from './graphCardContext';
import { graphCardHelpers } from './graphCardHelpers';
import { Chart } from '../chart/chart';
import { GraphCardChartLegend } from './graphCardChartLegend';
import { GraphCardChartTooltip } from './graphCardChartTooltip';
import { GraphCardChartTitleTooltip } from './graphCardChartTitleTooltip';
import { RHSM_API_QUERY_SET_TYPES } from '../../services/rhsm/rhsmConstants';
import { MinHeight } from '../minHeight/minHeight';
import { Loader } from '../loader/loader';
import { translate } from '../i18n/i18n';

/**
 * A chart/graph card.
 *
 * @param {object} props
 * @param {Function} props.t
 * @param {Function} props.useGetGraphData
 * @param {Function} props.useGraphCardContext
 * @param {Function} props.useProduct
 * @param {Function} props.useProductGraphTallyQuery
 * @returns {Node}
 */
const GraphCardChart = ({
  t,
  useGetGraphData: useAliasGetGraphData,
  useGraphCardContext: useAliasGraphCardContext,
  useProduct: useAliasProduct,
  useProductGraphTallyQuery: useAliasProductGraphTallyQuery
}) => {
  const { productId } = useAliasProduct();
  const { settings = {} } = useAliasGraphCardContext();
  // const { actionDisplay, metric, metrics = [] } = settings;
  const { actionDisplay, metric } = settings;

  const { [RHSM_API_QUERY_SET_TYPES.GRANULARITY]: granularity } = useAliasProductGraphTallyQuery();
  const { pending, error, data = {}, dataSets = [] } = useAliasGetGraphData();
  /*
  const { pending, error, data = {}, dataSets = [] } = useAliasMetricsSelector();
  const getGraphTally = useAliasGetGraphTally();

  useShallowCompareEffect(() => {
    getGraphTally(metrics.map(({ id: metricId }) => ({ id: productId, metric: metricId })));
  }, [metrics, productId, getGraphTally]);
  */

  const standaloneMetricId = (metric?.id && `_${metric?.id}`) || '';
  let updatedActionDisplay = null;

  if (typeof actionDisplay === 'function') {
    updatedActionDisplay = actionDisplay({ data: { ...data } });
  }

  return (
    <Card className="curiosity-usage-graph">
      <MinHeight key="headerMinHeight">
        <CardHeader>
          <CardTitle>
            <Title headingLevel="h2" size="lg">
              {t(`curiosity-graph.cardHeading${standaloneMetricId}`, { context: productId })}
              <GraphCardChartTitleTooltip />
            </Title>
          </CardTitle>
          {updatedActionDisplay && (
            <CardActions className={(error && 'blur') || ''}>{updatedActionDisplay}</CardActions>
          )}
        </CardHeader>
      </MinHeight>
      <MinHeight key="bodyMinHeight">
        <CardBody>
          <div className={(error && 'blur') || (pending && 'fadein') || ''}>
            {pending && <Loader variant="graph" />}
            {!pending && (
              <Chart
                {...graphCardHelpers.generateExtendedChartSettings({ settings, granularity })}
                dataSets={dataSets}
                chartLegend={({ chart, datum }) => <GraphCardChartLegend chart={chart} datum={datum} />}
                chartTooltip={({ datum }) => <GraphCardChartTooltip datum={datum} />}
              />
            )}
          </div>
        </CardBody>
      </MinHeight>
    </Card>
  );
};

/**
 * Prop types.
 *
 * @type {{useGraphCardContext: Function, useProduct: Function, useProductGraphTallyQuery: Function,
 *     t: Function, useGetGraphData: Function}}
 */
GraphCardChart.propTypes = {
  t: PropTypes.func,
  useGetGraphData: PropTypes.func,
  // useGetGraphTally: PropTypes.func,
  useGraphCardContext: PropTypes.func,
  // useMetricsSelector: PropTypes.func,
  useProduct: PropTypes.func,
  useProductGraphTallyQuery: PropTypes.func
};

/**
 * Default props.
 *
 * @type {{useGraphCardContext: Function, useProduct: Function, useProductGraphTallyQuery: Function,
 *     t: translate, useGetGraphData: Function}}
 */
GraphCardChart.defaultProps = {
  t: translate,
  useGetGraphData: useGetMetrics,
  // useGetGraphTally,
  useGraphCardContext,
  useProduct,
  useProductGraphTallyQuery
};

export { GraphCardChart as default, GraphCardChart };
