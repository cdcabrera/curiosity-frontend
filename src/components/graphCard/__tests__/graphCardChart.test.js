import React from 'react';
import { GraphCardChart } from '../graphCardChart';

describe('GraphCardChart Component', () => {
  it('should render a default component', () => {
    const props = {};
    const component = renderComponent(<GraphCardChart {...props} />);

    expect(component).toMatchSnapshot('default');
  });

  it('should handle API response states', () => {
    const props = {
      useGetMetrics: () => ({ pending: true, error: false, fulfilled: false, dataSets: [] })
    };

    const componentPending = renderComponent(<GraphCardChart {...props} />);
    expect(componentPending).toMatchSnapshot('pending');

    props.useGetMetrics = () => ({ pending: false, error: true, fulfilled: false, dataSets: [] });
    const componentError = renderComponent(<GraphCardChart {...props} />);
    expect(componentError).toMatchSnapshot('error');

    props.useGetMetrics = () => ({ pending: false, error: false, fulfilled: true, dataSets: [] });
    const componentFulfilled = renderComponent(<GraphCardChart {...props} />);
    expect(componentFulfilled).toMatchSnapshot('fulfilled');
  });
});
