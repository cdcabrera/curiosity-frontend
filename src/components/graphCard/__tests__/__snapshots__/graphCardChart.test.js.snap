// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphCardChart Component should handle API response states: error 1`] = `
<graphcardchart
  props="{
  "children": []
}"
>
  <article
    class="pf-c-card pf-m-plain curiosity-usage-graph"
    data-ouia-component-id="OUIA-Generated-Card-3"
    data-ouia-component-type="PF4/Card"
    data-ouia-safe="true"
    id=""
  >
    <div
      class="pf-c-card__header"
    >
      <div
        class="pf-c-card__title"
      >
        <h2
          class="pf-c-title pf-m-lg"
          data-ouia-component-id="OUIA-Generated-Title-3"
          data-ouia-component-type="PF4/Title"
          data-ouia-safe="true"
        >
          t(curiosity-graph.cardHeading, {})
        </h2>
      </div>
    </div>
    <div
      class="curiosity-minheight"
    >
      <div>
        <div
          class="pf-c-card__body"
        >
          <div
            class="blur"
          >
            <div
              class="curiosity-chartarea uxui-curiosity__modal uxui-curiosity__modal--loading"
              data-test="curiosity-chartarea"
            />
          </div>
        </div>
      </div>
    </div>
  </article>
</graphcardchart>
`;

exports[`GraphCardChart Component should handle API response states: fulfilled 1`] = `
<graphcardchart
  props="{
  "children": []
}"
>
  <article
    class="pf-c-card pf-m-plain curiosity-usage-graph"
    data-ouia-component-id="OUIA-Generated-Card-4"
    data-ouia-component-type="PF4/Card"
    data-ouia-safe="true"
    id=""
  >
    <div
      class="pf-c-card__header"
    >
      <div
        class="pf-c-card__title"
      >
        <h2
          class="pf-c-title pf-m-lg"
          data-ouia-component-id="OUIA-Generated-Title-4"
          data-ouia-component-type="PF4/Title"
          data-ouia-safe="true"
        >
          t(curiosity-graph.cardHeading, {})
        </h2>
      </div>
    </div>
    <div
      class="curiosity-minheight"
    >
      <div>
        <div
          class="pf-c-card__body"
        >
          <div
            class=""
          >
            <div
              class="curiosity-chartarea uxui-curiosity__modal uxui-curiosity__modal--loading"
              data-test="curiosity-chartarea"
            />
          </div>
        </div>
      </div>
    </div>
  </article>
</graphcardchart>
`;

exports[`GraphCardChart Component should handle API response states: pending 1`] = `
<graphcardchart
  props="{
  "children": []
}"
>
  <article
    class="pf-c-card pf-m-plain curiosity-usage-graph"
    data-ouia-component-id="OUIA-Generated-Card-2"
    data-ouia-component-type="PF4/Card"
    data-ouia-safe="true"
    id=""
  >
    <div
      class="pf-c-card__header"
    >
      <div
        class="pf-c-card__title"
      >
        <h2
          class="pf-c-title pf-m-lg"
          data-ouia-component-id="OUIA-Generated-Title-2"
          data-ouia-component-type="PF4/Title"
          data-ouia-safe="true"
        >
          t(curiosity-graph.cardHeading, {})
        </h2>
      </div>
    </div>
    <div
      class="curiosity-minheight"
    >
      <div>
        <div
          class="pf-c-card__body"
        >
          <div
            class="fadein"
          >
            <div
              class="curiosity-skeleton-container curiosity-skeleton-container-chart"
            >
              <div
                class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__lg"
              >
                <span
                  class="pf-u-screen-reader"
                />
              </div>
              <div
                class="pf-c-skeleton ins-c-skeleton ins-c-skeleton__md"
              >
                <span
                  class="pf-u-screen-reader"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </article>
</graphcardchart>
`;

exports[`GraphCardChart Component should render a default component: default 1`] = `
<graphcardchart
  props="{
  "children": []
}"
>
  <article
    class="pf-c-card pf-m-plain curiosity-usage-graph"
    data-ouia-component-id="OUIA-Generated-Card-1"
    data-ouia-component-type="PF4/Card"
    data-ouia-safe="true"
    id=""
  >
    <div
      class="pf-c-card__header"
    >
      <div
        class="pf-c-card__title"
      >
        <h2
          class="pf-c-title pf-m-lg"
          data-ouia-component-id="OUIA-Generated-Title-1"
          data-ouia-component-type="PF4/Title"
          data-ouia-safe="true"
        >
          t(curiosity-graph.cardHeading, {})
        </h2>
      </div>
    </div>
    <div
      class="curiosity-minheight"
    >
      <div>
        <div
          class="pf-c-card__body"
        >
          <div
            class=""
          >
            <div
              class="curiosity-chartarea uxui-curiosity__modal uxui-curiosity__modal--loading"
              data-test="curiosity-chartarea"
            />
          </div>
        </div>
      </div>
    </div>
  </article>
</graphcardchart>
`;
